### ISSUE TITLE
Faster

### ISSUE THREAD

**Issue by taylortrimble** [teslaworksumn/teslaworks.net#120]
We need to speed our puppy up! Most of it is my fault. Things to fix:
- [ ] Resize images
- [ ] Requalitize images
- [x] Investigate latency
- [x] Investigate `@2x` lazy loading of images
- [x] Investigate Etag or Last-Modified caching on the html pages

**Comment by taylortrimble** [comment]
Today's investigation yields:
- We need to focus on latency vs bandwidth
  - In a worst-case load (8.66s for the front page), 6.10s of our load time comes from latency
  - Latency-bound means we aren't in horrible shape!
  - After about 6 requests, our latency shoots up
  - Latency is usually quite hard to fix
- File size can be an issue too: much less of one, but "next"
  - This one 1.1MB file keeps loading in 7.5ms!! Srsly.That or 2+s. :wink: Just a fun anomaly.
  - Our logo is much, much smaller as a PNG and takes almost no time to load. Makes sense, and it has always been one
  - Our longest downloads are 700+KB images, those can come ways down. Prob. save about 1s, which is **not** small, so that's good
- De-retinizing
  - The JPEGs don't need to be retina! I've seen them on my phone, and our nasty scaled-up JPEGs are acceptable.
  - Retina usually only looks bad for sprites and logos anyway: I'll leave the logos retina PNGs and scale the rest
  - Lazy-loading of retina resources is achievable, but the solutions aren't ideal:
    1. Replace `img` tags with `div`s with CSS background images, which is horrible and we lose bootstrap styling
    2. Set a cookie, custom static serve. Probably worth experimenting with, but I bet we balls out on our typical latency
    3. Swapping the images at runtime, aka triple bandwidth on phones. Not happening. The use case of phones is in-person demos of the site, which **cannot** be slow (and it's probably over LTE/3G to boot!).

**Comment by taylortrimble** [comment]
Resolved: no cacheing the HTTP. No gains.

**Comment by taylortrimble** [comment]
> Set a cookie, custom static serve. Probably worth experimenting with, but I bet we balls out on our typical latency.

If Nginx static serving is successful, we won't do this, which rules out any nice options for keeping retina. Let's skimp out on the JPEGs and keep the PNGs beautiful as the course of action.

**Comment by mplewis** [comment]
Look into ImageOptim and ImageAlpha. Image downsizing is somewhat of an art
when you start cutting alpha channels.

Let's provide retina. I do notice a serious difference, and I believe other
designers do too. We can do the @2X lazyload thing. If it's good enough for
Apple, it's good enough for us.

You're right, though. We need to serve small images for mobile. Looksy:
http://adaptive-images.com/

**Comment by bendemaree** [comment]
I'm suspecting this is a server issue. I'm seeing latency of ~45ms and a load time of 40-50ms...that's awesome for Flask. Based on a page weight with no cache hits of 4.3M you're doing well; I have a rendered page with all assets in <2s. Guessing these numbers are in direct relation to the change o' server last night; don't need to pull your hair out on minute optimizations if a server is at fault. :smiley:

If you are seeing load time spikes after repeat requests, and they are due to the app, always drop in Memcached. I'd be happy to PR [Flask-Cache](http://pythonhosted.org/Flask-Cache/) in here if you think that's something you want to go forward with.

Do keep HiDPI. Good UX.

**Comment by bendemaree** [comment]
I should revise; I see you're instantiating the controller outside the request loop to keep your projects loaded from JSON in-memory, which is a good technique; nice work! There's an edge case here though where none of the projects qualify as `past_projects` and therefore `self.past_projects` evaluates to False on every request, and the controller dutifully goes to load those projects as if it's the first load of the app.

TL;DR you're unintentionally hitting disk on every request. Want a PR?

**Comment by taylortrimble** [comment]
Right you are! Go for it!

**Comment by taylortrimble** [comment]
Now your numbers!

> I'm seeing latency of ~45ms and a load time of 40-50ms…that's awesome for Flask.

I've _never_ seen that yet! Our server you saw that? :heart_eyes: I don't even know: Here's what I typically see:

#### Good case (and with cache hits)

![good-case](https://f.cloud.github.com/assets/1058478/1139268/f0af0d7e-1c87-11e3-986d-0da093c23d65.png)

#### Bad case (no cache hits)

![bad-case](https://f.cloud.github.com/assets/1058478/1139269/f72969e2-1c87-11e3-9e16-6fcd6d2f3d29.png)

Even in the good case, I've never seen anything less than 100ms of latency for the non-HTML resources. The bad case also exhibits that awful ramp-up on later request latency when there's no cache hits.

### Change of server

Actually seeing some performance **degredation** after the switch, though not in the typical case. Startup time can be pretty bad though:

![screen shot 2013-09-13 at 3 10 12 am](https://f.cloud.github.com/assets/1058478/1140050/d0a774d4-1c97-11e3-93d5-32c6c0a7270c.png)

That's a bad page load time. I don't think that's super common, but I'll keep an eye out.

### Memcached

Probably not necessary yet: our load latency is for static serving.

### HiDPI

We'll see… I just don't see an improvement with those JPEGs. I can test when I resize/requal though. @mplewis: good image tools!

**Comment by taylortrimble** [comment]
### New Server!

So, corresponding to how **bad** the new server was when we first fire it up (1.5 **minute** page loads) its warm up speed is something to like.

Old no-cache latency: > 2s on repeat request load, bad cases > 4s.
New no-cache latency: usually less than 1s in the worst case. We don't see the fastest latencies being as fast anymore (order of 10ms -> order of 100ms) but overall a nice improvement.

### Refresh vs revisit

I'm noticing a difference between when I refresh my page and when I revisit it. On a refresh, we usually have around 600ms with cacheing. With a revisit, 350ms with almost zero (1.2ms) static resource latency! Now we've moved from a fastest on the order of 100ms to the order of 1ms.

Any explanation? Also, I'm starting to like this new server again. :)

**Comment by bendemaree** [comment]
I can come back to this in detail in a bit but this is what I'm seeing. I put disproportionate emphasis on the first request because if the page can start to render after that downloads, the UX is improved. And in that realm we're doing well.

![screen shot 2013-09-13 at 1 33 26 pm](https://f.cloud.github.com/assets/1188702/1140920/aae2d8ea-1ca8-11e3-83a7-e9a5143b1168.png)

**Comment by taylortrimble** [comment]
Yeah that HTML comes in like lightning! That's on par with what I'm experiencing now too. I think that if nginx helps us cut down on that static latency and we properly size our images (no more 600KB+) we'll be FUKKEN SMOKIN IT.

**Comment by mplewis** [comment]
Looks relevant:
http://stackoverflow.com/questions/2451739/django-serving-static-files-through-nginx

**Comment by taylortrimble** [comment]
For sure! You want to proceed together with testing and implementing?
